### PR TITLE
fixing error in _ion_mass

### PR DESCRIPTION
**Pull request by claytonstrawn**
I was using total gas masses to see overall how much oxygen is in each species, and it looked like my total O mass was actually the same as the total hydrogen mass. That did not seem right, and I think the problem came from the _ion_mass function, who's return statement really doesn't make sense as is.

My fix might not be the best, because I'm not sure if `data['gas','cell_volume']` is general enough for me to be able to use mass=density*volume, but that's the most natural way I see that lets you generate the fields in the same order as they do now.

Otherwise you'd need to step through the different ways metals can be stored (`nuclei_mass_density`, `species_metallicity`, `solar_abundance`, etc. again)

### CODE REVIEW

**Comment by chummels** [comment]
Hi @claytonstrawn .  It appears that yes, there is a problem with the `_ion_mass()` function in the `sph-viz` branch, but your fix will not address the problem.  `cell_volume` is a field that is only defined for grid-based simulations, which is the reason that the field definition was modified when we moved to the demeshened version of the code.  

One identified issue in yt is to explicitly define a volume field in the yt-4.0 version of the code for particle-based datasets (https://github.com/yt-project/yt/issues/1494).  Once this issue is addressed, it will provide a clear solution to the problem you have identified.  But your correction here won't work, unfortunately.  

Until then, we could try re-defining the field to be a piecewise function, similar to what was done in the `master` branch:  https://github.com/trident-project/trident/blob/master/trident/ion_balance.py#L867

**Comment by claytonstrawn** [comment]
We could also do the same thing as _number_density does and just switch places where it looks for `[ftype,"density"]` to `[ftype,"mass"]`

**Comment by brittonsmith** [comment]
@chummels, @claytonstrawn I think Clayton's last comment is likely the way to go. All SPH frontends should have a native mass field for each particle type, so I think it would just be a matter of going with `(ftype, 'mass')`. Presumably, for grid data, `ftype` is always "gas", so I don't see why that won't just work.

**Comment by claytonstrawn** [comment]
@chummels @brittonsmith  okay, I added a new commit which does this. I just directly rewrote _number_density it in terms of mass fields and changed the names/some variables. I'm not sure how often, if ever, things like `H_nuclei_mass` fields exist, but it will just skip them if they aren't.

File: trident/ion_balance.py [hunk 1]
```diff
@@ -686,23 +686,50 @@ def add_ion_mass_field(atom, ion, ds, ftype="gas",
...
+    atom = field_name.split("_")[0]
     prefix = field_name.split("_mass")[0]
     suffix = field_name.split("_mass")[-1]
     fraction_field_name = "%s_ion_fraction%s" % (prefix, suffix)
-    return data[ftype, fraction_field_name] * data["gas", "mass"]
+
+    # try the atom-specific density field first
+    nuclei_field = "%s_nuclei_mass_density" % atom
+    if (ftype, nuclei_field) in data.ds.field_info:
+        return data[fraction_field_name] * \
```
Comment (brittonsmith) [inline trident/ion_balance.py]: I think this just needs `ftype`.
```suggestion
        return data[ftype, fraction_field_name] * \
```

File: trident/ion_balance.py [hunk 2]
```diff
@@ -686,23 +686,50 @@ def add_ion_mass_field(atom, ion, ds, ftype="gas",
...
+          data[(ftype, nuclei_field)]
+
+    # try the species metallicity
+    metallicity_field = "%s_metallicity" % atom
+    if (ftype, metallicity_field) in data.ds.field_info:
+        return data[fraction_field_name] * \
```
Comment (brittonsmith) [inline trident/ion_balance.py]: Here, too.
```suggestion
        return data[ftype, fraction_field_name] * \
```

File: trident/ion_balance.py [hunk 3]
```diff
@@ -686,23 +686,50 @@ def add_ion_mass_field(atom, ion, ds, ftype="gas",
...
+          data[ftype, "mass"] * \
+          data[ftype, metallicity_field]
+    
+    if atom == 'H' or atom == 'He':
+        mass_fraction = solar_abundance[atom] * data[fraction_field_name]
```
Comment (brittonsmith) [inline trident/ion_balance.py]:
```suggestion
        mass_fraction = solar_abundance[atom] * data[ftype, fraction_field_name]
```

File: trident/ion_balance.py [hunk 4]
```diff
@@ -686,23 +686,50 @@ def add_ion_mass_field(atom, ion, ds, ftype="gas",
...
+    else:
+        mass_fraction = data.ds.quan(solar_abundance[atom], "1.0/Zsun") * \
+          data[ftype, fraction_field_name] * \
+          data[ftype, "metallicity"]
+    # convert to total mass
+    # use the on disk hydrogen mass if possible
+    if (ftype, "H_nuclei_mass") in data.ds.derived_field_list:
+        mass = mass_fraction * data[ftype, "H_nuclei_mass"]
+    else:
+        mass = mass_fraction * data[ftype, "mass"]
```
Comment (brittonsmith) [inline trident/ion_balance.py]: In the case where we don't have H nuclei density, it's probably a bit more accurate to assume a solar H mass fraction, so I would multiply this line by `H_mass_fraction`, defined up top as 0.76.

File: trident/ion_balance.py (hunk 3 above)
Comment (claytonstrawn) [reply trident/ion_balance.py]: Done! Though it uses that framework on `_ion_number_density` too, so we may want to also change that to use `ftype`

File: trident/ion_balance.py (hunk 4 above)
Comment (claytonstrawn) [reply trident/ion_balance.py]: Oh right, I didn't realize `to_nH` was not just `1/mh`. Fixed

File: trident/light_ray.py [hunk 5]
```diff
@@ -451,6 +451,10 @@ def make_light_ray(self, seed=None, periodic=True,
         ...                       use_peculiar_velocity=True)
 
         """
+
+        # force unitary units to exist if they don't already
+        self.code_length.in_units("unitary")
```
Comment (brittonsmith) [inline trident/light_ray.py]: `in_units` will return a copy, but not alter the units of the existing array. For that, you'll need `convert_to_units`.

Comment (brittonsmith) [reply trident/light_ray.py]: Actually, `self` here does not have a `code_length` attribute. What's the goal of this here?

Comment (claytonstrawn) [reply trident/light_ray.py]: Oh this was just me testing something. I didn't realize it had submitted this with the pull. Let me remove it.

**Comment by chummels** [comment]
I ran into the ion masses being wrong in demesh and it occurred to me that this never got merged.  What is the status of this PR?

**Comment by claytonstrawn** [comment]
@chummels I'm not sure. I think I screwed up the PR by doing a pull from upstream, and now it has contributions from @boydbre1 in here too. But my _ion_mass function seems to work ok.

**Comment by brittonsmith** [comment]
@claytonstrawn, we're having issues with some changes that went into yt. I'm working on that now. As soon as I get them cleared up I'll restart the tests and we can merge this is everything passes. I may have to ask you to pull in some trident changes, but I'll let you know.

**Comment by brittonsmith** [comment]
@claytonstrawn, I think the issue was that you pulled in some changes that had only gone into trident's `master` branch, but I think I have the conflicts all sorted out. It's probably better this way, though, because you've resolved some potential merger conflicts on your own.

Testing is now functional again, so if you  could just look this over one more time and make sure everything looks as you think it should, I'll go ahead an merge, assuming that the tests pass.

**Comment by claytonstrawn** [comment]
This looks good to me. _ion_mass is still fixed, that's the real content of this change (some other things were just to make the other field functions have the same style).

I would go ahead and merge it.

**Comment by brittonsmith** [comment]
Ok, great, thanks @claytonstrawn!

**Pull request merged**